File: play-services-core/src/main/java/org/microg/gms/rcs/RcsService.java
/*
 * SPDX-FileCopyrightText: 2026 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.gms.rcs;

import android.os.RemoteException;
import android.util.Log;

import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.common.internal.ConnectionInfo;
import com.google.android.gms.common.internal.GetServiceRequest;
import com.google.android.gms.common.internal.IGmsCallbacks;

import org.microg.gms.BaseService;
import org.microg.gms.common.GmsService;
import org.microg.gms.common.PackageUtils;

/**
 * RCS Service implementation.
 *
 * Handles the "com.google.android.gms.rcs.START" intent.
 * Returns a success code to allow Google Messages to proceed with RCS setup.
 */
public class RcsService extends BaseService {
    private static final String TAG = "GmsRcsService";

    public RcsService() {
        super(TAG, GmsService.RCS);
    }

    @Override
    public void handleServiceRequest(IGmsCallbacks callback, GetServiceRequest request, GmsService service) throws RemoteException {
        String packageName = PackageUtils.getAndCheckCallingPackage(this, request.packageName);
        if (packageName == null) {
            Log.w(TAG, "Missing or invalid calling package");
            return;
        }
        String callingVersion = RcsCallerPolicy.getPackageVersionSummary(this, packageName);

        Log.d(TAG, "handleServiceRequest from: " + packageName);
        Log.i("MicroGRcs", "svc189 bind caller=" + packageName + " version=" + callingVersion + " supportsConnectionInfo=" + request.supportsConnectionInfo);
        if (request.extras != null) {
            Log.d(TAG, "Request extras: " + request.extras);
        }

        // Return SUCCESS and our binder stub
        callback.onPostInitCompleteWithConnectionInfo(
            CommonStatusCodes.SUCCESS,
            new RcsServiceImpl(packageName).asBinder(),
            new ConnectionInfo()
        );
    }

    private static class RcsServiceImpl extends IRcsService.Stub {
        private final String packageName;

        public RcsServiceImpl(String packageName) {
            this.packageName = packageName;
        }
    }
}

File: play-services-core/src/main/kotlin/org/microg/gms/rcs/RcsService.kt
/*
 * SPDX-FileCopyrightText: 2026 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.gms.rcs

import android.os.Binder
import android.os.Parcel
import android.util.Log
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.internal.ConnectionInfo
import com.google.android.gms.common.internal.GetServiceRequest
import com.google.android.gms.common.internal.IGmsCallbacks
import org.microg.gms.BaseService
import org.microg.gms.common.GmsService
import org.microg.gms.common.PackageUtils
import org.microg.gms.utils.warnOnTransactionIssues

private const val TAG = "GmsRcsService"

/**
 * Google Messages binds [GmsService.RCS] (`com.google.android.gms.rcs.START`) during setup.
 * DummyService answers that bind with [com.google.android.gms.common.ConnectionResult.API_DISABLED].
 *
 * Unknown binder codes are logged via [warnOnTransactionIssues] so they can be mapped later.
 */
class RcsService : BaseService(TAG, GmsService.RCS) {
    override fun handleServiceRequest(callback: IGmsCallbacks, request: GetServiceRequest, service: GmsService) {
        val packageName = PackageUtils.getAndCheckCallingPackage(this, request.packageName)
        if (!PackageUtils.isGooglePackage(this, packageName)) {
            throw SecurityException("$packageName is not a Google package")
        }
        Log.d(TAG, "handleServiceRequest from $packageName")
        callback.onPostInitCompleteWithConnectionInfo(
            CommonStatusCodes.SUCCESS,
            RcsServiceImpl(),
            ConnectionInfo()
        )
    }
}

class RcsServiceImpl : Binder() {
    override fun onTransact(code: Int, data: Parcel, reply: Parcel?, flags: Int): Boolean =
        warnOnTransactionIssues(code, reply, flags, TAG) { super.onTransact(code, data, reply, flags) }
}
